# SiNPle - Simplified Inference of Novel Polymorphisms from Large coveragE

## Overview

`SiNPle` is a fast and effective software for variant calling based on a simplified Bayesian approach to compute the posterior probability that a variant is not generated by sequencing errors or PCR artefacts. It is particularly suited to low-frequency variant calling, as described in [Ferretti (2019)](https://www.mdpi.com/2073-4425/10/8/561/htm).
While calling heterozygous variants in samples with defined ploidy is relatively straightforward &mdash; for instance, the expected allele frequency is 50% or 100% in the diploid case &mdash; the underlying assumptions for low-frequency variant calling are very different and require a different statistical approach.
In the Bayesian model used by `SiNPle` we explicitly consider the limit in which ploidy tends to infinity, which is equivalent to a mixture with an infinite number of molecules, and thus no artificial constraints are imposed.

## Comparison with other callers

Some genotype callers, such as GATK, assume a ploidy of two. Others have hidden hardwired thresholds &mdash; for instance, a minimum number of reads needed to make a call &mdash; and hence miss a number of low-frequency variants. A detailed comparison with low-frequency variant callers LoFreq and VarScan2 performed on CirSeq data can be found in the published paper ([Ferretti, 2019](https://www.mdpi.com/2073-4425/10/8/561/htm)).

## Citation

If you use SiNPle, please cite

## Installation

To compile SiNPle, please install ocaml (http://www.ocaml.org/docs/install.html) and dune and run the following command:

  ./BUILD

## How to run it

This will compile the binary 'SiNPle' and 'SiNPle -h' would give information on all the options.
An example of generating SiNPle variant calls from the bamfile example.bam would be

  samtools mpileup -d 1000000 -a -A -B -Q 0 -x example.bam | SiNPle > variants.txt

## Command line syntax

```
This is the SiNPle variant calling program (version 0.8)
 (c) 2017-2019 Luca Ferretti, <luca.ferretti@gmail.com>
 (c) 2017-2019 Chandana Tennakoon, <drcyber@gmail.com>
 (c) 2017-2021 Paolo Ribeca, <paolo.ribeca@gmail.com>
```
Usage:
```
./SiNPle
```

Algorithmic parameters
| Option | Argument(s) | Effect | Note(s) |
|-|-|-|-|
| `-t`<br>`--theta` | _&lt;non\_negative\_float&gt;_ |  prior estimate of nucleotide diversity | <ins>default=<mark>_0\.001_</mark></ins> |
| `-T`<br>`--theta-indel` | _&lt;non\_negative\_float&gt;_ |  prior estimate of indel likelihood | <ins>default=<mark>_0\.0001_</mark></ins> |
| `-I`<br>`--quality-indel-short` | _&lt;non\_negative\_integer&gt;_ |  prior Phred\-scaled quality for indels of length 1 | <ins>default=<mark>_35_</mark></ins> |
| `-L`<br>`--quality-indel-long` | _&lt;non\_negative\_integer&gt;_ |  prior Phred\-scaled quality for indels of length &gt;1 | <ins>default=<mark>_45_</mark></ins> |
| `-p`<br>`--pcr-error-rate` | _&lt;non\_negative\_float&gt;_ |  prior estimate of error rate for PCR\-generated substitutions | <ins>default=<mark>_5e\-07_</mark></ins> |
| `-P`<br>`--pcr-error-rate-indel` | _&lt;non\_negative\_float&gt;_ |  prior estimate of error rate for PCR\-generated indels | <ins>default=<mark>_5e\-08_</mark></ins> |
| `--error-rate-substitution` | _&lt;non\_negative\_float&gt;_ |  prior estimate of error rate for sequencing\-generated substitutions | <ins>default=<mark>_0\.0001_</mark></ins> |
| `--error-rate-indel-short` | _&lt;non\_negative\_float&gt;_ |  prior estimate of error rate for sequencing\-generated indels of length 1 | <ins>default=<mark>_1e\-05_</mark></ins> |
| `--error-rate-indel-long` | _&lt;non\_negative\_float&gt;_ |  prior estimate of error rate for sequencing\-generated indels of length &gt;1 | <ins>default=<mark>_1e\-06_</mark></ins> |
| `-s`<br>`-S`<br>`--strandedness` | _forward&#124;reverse&#124;both_ |  strands to be taken into account for counts | <ins>default=<mark>_both_</mark></ins> |

Input/Output
| Option | Argument(s) | Effect | Note(s) |
|-|-|-|-|
| `-i`<br>`--input` | _&lt;input\_file&gt;_ |  name of input file \(in mpileup format\) | <ins>default=<mark>_&lt;stdin&gt;_</mark></ins> |
| `-o`<br>`--output` | _&lt;output\_file&gt;_ |  name of output file | <ins>default=<mark>_&lt;stdout&gt;_</mark></ins> |

Miscellaneous
| Option | Argument(s) | Effect | Note(s) |
|-|-|-|-|
| `-v`<br>`--version` |  |  print version and exit |  |
| `-h`<br>`--help` |  |  print syntax and exit |  |

